### PR TITLE
Add shamir secret sharing csharp wrapper

### PR DESCRIPTION
**Pull request by MathieuMorrissette**
Code from branch shamir-wrappers (Squash all the commits into a single one)

### CODE REVIEW

File: wrappers/csharp/Managed.cs [hunk 1]
```diff
@@ -862,6 +863,43 @@ public static string DecryptWithPasswordAsString(string b64data, string password
...
+                int sharesLength = shares[0].Length;
+                int secretLength = (int)Native.JoinSharesSizeNative((UIntPtr)sharesLength);
+                byte[] secret = new byte[secretLength];
+                byte[][] sharesTemp = new byte[nbShares][];
+                Array.Copy(shares, sharesTemp, shares.Length);
+                IntPtr[] pointers = GetArrayReferences(ref shares, nbShares, secretLength);
+                Array.Copy(sharesTemp, shares, sharesTemp.Length);
+                long result = Native.JoinSharesNative((UIntPtr)nbShares, (UIntPtr)sharesLength, pointers, secret, (UIntPtr)secretLength);
+
+                return secret;
```
Comment (pdugre) [inline wrappers/csharp/Managed.cs]: 1. You should free the pointers after you're done with it.
2. The native errors aren't properly handled

File: wrappers/csharp/Managed.cs [hunk 2]
```diff
@@ -910,6 +948,29 @@ public static string DecryptWithPasswordAsUtf8String(string b64data, string pass
...
+
+                return shares;
+            }
+            catch (DevolutionsCryptoException ex)
+            {
+                if (ex.NativeError != null)
+                {
+                    Utils.HandleError((long)ex.NativeError);
+                }
+            }
```
Comment (pdugre) [inline wrappers/csharp/Managed.cs]: 1. You should free the pointers after you're done with it.
2. The native errors aren't properly handled

File: wrappers/csharp/Managed.cs (hunk 1 above)
Comment (cbienaime) [reply wrappers/csharp/Managed.cs]: Pointers are released inside the GetArrayReferences and InitializeArray methods. In the example, It is the handles [i] .Free (), which must be done

Comment (pdugre) [reply wrappers/csharp/Managed.cs]: Free() call should still be done *after* they are used, which means after the Native.JoinSharesNative() call

File: wrappers/csharp/Managed.cs [hunk 3]
```diff
@@ -1,16 +1,22 @@
 namespace Devolutions.Cryptography
 {
     using System;
+    using System.Linq;
```
Comment (MathieuMorrissette) [inline wrappers/csharp/Managed.cs]: Is Linq required just for the .All() in the method SharesLengthAreValid below? If so, a simple for loop would be much faster.

Comment (cbienaime) [reply wrappers/csharp/Managed.cs]: Fix

File: wrappers/csharp/Managed.cs [hunk 4]
```diff
@@ -862,6 +861,55 @@ public static string DecryptWithPasswordAsString(string b64data, string password
...
         }
 
+        /// <summary>
+        /// Join multiple shares to regenerate a shared secret.
+        /// </summary>
+        /// <param name="shares">This is a 2-dimensional array representing the shares.</param>
+        /// <returns>The output buffer with the shared secret.</returns>
+        public static byte[] JoinShares(byte[][] shares)
+        {
+            if (shares == null || shares.Length == 0 || !SharesLengthAreValid(shares))
```
Comment (pdugre) [inline wrappers/csharp/Managed.cs]: shares == null and shares.Length == 0 is validated 2 times, so you can remove the checks here and leave only !SharesLengthAreValid(shares)

File: wrappers/csharp/Managed.cs [hunk 5]
```diff
@@ -1018,5 +1104,56 @@ public static Guid GenerateAPIKey()
...
+                pointers[i] = handles[i].AddrOfPinnedObject();
+            }
+
+            for (int i = 0; i < handles.Length; i++)
+            {
+                handles[i].Free();
+            }
+
+            return pointers;
+        }
```
Comment (pdugre) [inline wrappers/csharp/Managed.cs]: Unused method

File: wrappers/csharp/Native.cs [hunk 6]
```diff
@@ -86,6 +86,46 @@ public static bool VerifyPassword(byte[] password, byte[] hash)
...
+        public static long GenerateSharedKeySizeNative(UIntPtr secretLength)
+        {
+            if (Environment.Is64BitProcess)
+            {
+                return GenerateSharedKeySize64(secretLength);
+            }
+
+            return GenerateSharedKeySize86(secretLength);
+        }
+
```
Comment (MathieuMorrissette) [inline wrappers/csharp/Native.cs]: Do not expose the methods to the lib users ( Make them internal )

Comment (cbienaime) [reply wrappers/csharp/Native.cs]: Done

File: wrappers/csharp/Managed.cs [hunk 7]
```diff
@@ -862,6 +861,55 @@ public static string DecryptWithPasswordAsString(string b64data, string password
...
+        /// <param name="shares">This is a 2-dimensional array representing the shares.</param>
+        /// <returns>The output buffer with the shared secret.</returns>
+        public static byte[] JoinShares(byte[][] shares)
+        {
+            if (!SharesLengthAreValid(shares))
+            {
+                throw new DevolutionsCryptoException(ManagedError.InvalidParameter);
+            }
+
+            int nbShares = shares.Length;
```
Comment (pdugre) [inline wrappers/csharp/Managed.cs]: Disable the warning here:
```
#pragma warning disable CA1062 // nullcheck is already done in SharesLengthAreValid
int nbShares = shares.Length;
#pragma warning restore CA1062 // nullcheck is already done in SharesLengthAreValid
```

Comment (cbienaime) [reply wrappers/csharp/Managed.cs]: Done

**Pull request merged**